assets/arc-0052/contextual.api.crypto.ts

```diff
@@ -0,0 +1,292 @@
+            // const nodeCC: Uint8Array = derived.subarray(64, 96)
+
+            // // [Public][ChainCode]
+            // const extPub: Uint8Array = new Uint8Array(Buffer.concat([nodePublic, nodeCC]))
+            // const publicKey: Uint8Array = deriveChildNodePublic(extPub, bip44Path[4]).subarray(0, 32)
+
+            derived = deriveChildNodePrivate(derived, bip44Path[4])
+
+        const scalar = derived.subarray(0, 32) // scalar == pvtKey
+        return isPrivate ? scalar : crypto_scalarmult_ed25519_base_noclamp(scalar)
+    }
+
+    /**
+     * 
+     * 
+     * @param context - context of the key (i.e Address, Identity)
+     * @param account - account number. This value will be hardened as part of BIP44
+     * @param keyIndex - key index. This value will be a SOFT derivation as part of BIP44.
+     * @returns - public key 32 bytes
+     */
+    async keyGen(context: KeyContext, account:number, keyIndex: number): Promise<Uint8Array> {
+        await ready // libsodium
+
+        const rootKey: Uint8Array = fromSeed(this.seed)
+        const bip44Path: number[] = GetBIP44PathFromContext(context, account, keyIndex)
+
+        return await this.deriveKey(rootKey, bip44Path, false)
+    }
+
+    /**
+     * Ref: https://datatracker.ietf.org/doc/html/rfc8032#section-5.1.6
+     * 
+     *  Edwards-Curve Digital Signature Algorithm (EdDSA)
+     * 
+     * @param context - context of the key (i.e Address, Identity)
+     * @param account - account number. This value will be hardened as part of BIP44
+     * @param keyIndex - key index. This value will be a SOFT derivation as part of BIP44.
+     * @param data - data to be signed in raw bytes
+     * @param metadata - metadata object that describes how `data` was encoded and what schema to use to validate against
+     * 
+     * @returns - signature holding R and S, totally 64 bytes
+     * */ 
+    async signData(context: KeyContext, account: number, keyIndex: number, data: Uint8Array, metadata: SignMetadata): Promise<Uint8Array> {
+        // validate data
+        const result: boolean | Error = this.validateData(data, metadata)
+        
+        if (result instanceof Error) { // decoding errors
+            throw result
+        }
+
+        if (!result) { // failed schema validation
+            throw ERROR_BAD_DATA
+        }
+        
+        await ready // libsodium
+
+        const rootKey: Uint8Array = fromSeed(this.seed)
+        const bip44Path: number[] = GetBIP44PathFromContext(context, account, keyIndex)
+        const raw: Uint8Array = await this.deriveKey(rootKey, bip44Path, true)
+
+        const scalar = raw.slice(0, 32);
+        const c = raw.slice(32, 64);
+
+        // \(1): pubKey = scalar * G (base point, no clamp)
+        const publicKey = crypto_scalarmult_ed25519_base_noclamp(scalar);
+
+        // \(2): h = hash(c + msg) mod q
+        const hash: bigint = Buffer.from(crypto_hash_sha512(Buffer.concat([c, data]))).readBigInt64LE()
```

Comment:
I'm not sure about this line here. 

I called signData in the following way:

const message = Buffer.from("Hello World")
const signature: Uint8Array = await cryptoService.signData(KeyContext.Address, 0, 0, message, { encoding: Encoding.NONE, schema: {} })

When I print out the hash it looks like this:

<img width="331" alt="Screenshot 2024-02-21 at 10 29 34" src="https://github.com/algorandfoundation/ARCs/assets/83883690/e567d2c5-6f00-4d05-a998-d543dd298a5d">

And the calculated r looks like this:
<img width="508" alt="Screenshot 2024-02-21 at 10 30 12" src="https://github.com/algorandfoundation/ARCs/assets/83883690/e9ad565e-5ce7-4ebc-a238-99a1163d6c7c">

To me I think the hash looks strange, I would expect it to have the full length, and then we simply do a r = scalar_reduce(hash) to get it down under q with some reversal of bytes to get it into Little Endian mode.

Reply:
Updated